csharp/ql/src/Telemetry/ExternalApiSpecific.qll

```diff
@@ -0,0 +1,80 @@
+/** Provides classes and predicates related to handling APIs from external libraries. */
+
+import csharp
+import semmle.code.csharp.dataflow.DataFlow
+import semmle.code.csharp.dataflow.ExternalFlow
+import semmle.code.csharp.dataflow.FlowSummary
+import semmle.code.csharp.dataflow.TaintTracking
+import semmle.code.csharp.security.dataflow.flowsources.Remote
+private import semmle.code.csharp.dataflow.internal.DataFlowPrivate
+private import semmle.code.csharp.dataflow.internal.TaintTrackingPrivate
+
+/**
+ * An external API from either the C# Standard Library or a 3rd party library.
+ */
+class ExternalApiSpecific extends Callable {
+  ExternalApiSpecific() { this.fromLibrary() }
+
+  /**
+   * Gets the unbound type, name and parameter types of this API.
+   */
+  private string getSignature() {
+    result =
+      this.getDeclaringType().getUnboundDeclaration() + "." + this.getName() + "(" +
+        this.parameterTypesToString() + ")"
+  }
+
+  /**
+   * Gets the namespace of this API.
+   */
+  private string getNamespace() { result = this.getDeclaringType().getNamespace().toString() }
+
+  /**
+   * Gets the assembly file name containing this API.
+   */
+  private string getAssembly() { result = this.getFile().getBaseName() }
+
+  /**
+   * Gets the assembly file name and namespace of this API.
+   */
+  string getInfoPrefix() { result = this.getAssembly() + "#" + this.getNamespace() }
+
+  /**
+   * Gets the assembly file name, namespace and signature of this API.
+   */
+  string getInfo() { result = this.getInfoPrefix() + "#" + this.getSignature() }
+
+  /** Gets a node that is an input to a call to this API. */
+  DataFlow::Node getAnInput() {
+    exists(Call call | call.getTarget().getUnboundDeclaration() = this |
+      result.asExpr() = call.getAnArgument()
+    )
+    or
+    result.(ArgumentNode).getCall().getEnclosingCallable() = this
+  }
+
+  /** Gets a node that is an output from a call to this API. */
+  DataFlow::Node getAnOutput() {
+    exists(Call call | call.getTarget().getUnboundDeclaration() = this | result.asExpr() = call)
+    or
+    result.(PostUpdateNode).getPreUpdateNode().(ArgumentNode).getCall().getEnclosingCallable() =
+      this
+  }
```

Comment:
This will not include `out`/`ref` outputs. Something like this perhaps, which should also handle post-update nodes:
```ql
private import DataFlowImplCommon

  DataFlow::Node getAnOutput() {
    exists(NonDelegateDataFlowCall call, ReturnKindExt ret |
      this = [call.getDispatchCall().getADynamicTarget(), call.getDispatchCall().getAStaticTarget()).getUnboundDeclaration() and
      result = ret.getAnOutNode(call)
    )
  }
```

Reply:
Same comment as above.